cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -186,34 +209,96 @@ func generateMountsModeRecover(recoverySystem string) error {
+	if err != nil {
+		return err
+	}
+
+	// find the partition on that disk with the label
+	partuuid, err := disk.FindMatchingPartitionUUID(label)
+	if err != nil {
+		return err
+	}
+
+	// mount that and return
+	fmt.Fprintf(stdout, "/dev/disk/by-partuuid/%s %s\n", partuuid, target)
+	return nil
+}
+
 func generateMountsModeRun() error {
-	// 1.1 always ensure basic partitions are mounted
-	for _, d := range []string{boot.InitramfsUbuntuSeedDir, boot.InitramfsUbuntuBootDir} {
-		isMounted, err := osutilIsMounted(d)
-		if err != nil {
-			return err
-		}
-		if !isMounted {
-			fmt.Fprintf(stdout, "/dev/disk/by-label/%s %s\n", filepath.Base(d), d)
-		}
+	// 1.1 mount ubuntu-boot first, using the LoaderDevicePartUUID to identify
+	//     the partuiid partition that should be mounted
+	//     in run mode, we should be getting our kernel from ubuntu-boot
+
+	isMounted, err := osutilIsMounted(boot.InitramfsUbuntuBootDir)
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		return selectPartitionMatchingKernelDisk(boot.InitramfsUbuntuBootDir, "ubuntu-boot")
+	}
+
+	// 1.2 mount ubuntu-seed partition from same disk as ubuntu-boot
+	isMounted, err = osutilIsMounted(boot.InitramfsUbuntuSeedDir)
+	if err != nil {
+		return err
+	}
+	if !isMounted {
+		// get the disk that the ubuntu-boot mount point comes from, as that
+		// should have been where we booted the kernel from in run mode
+		return mountPartitionLabelFromSameDiskAsMountPoint(
+			boot.InitramfsUbuntuBootDir, // existing mountpoint
+			"ubuntu-seed",               // label to mount
+			boot.InitramfsUbuntuSeedDir, // destination mountpoint
+		)
+	}
+
+	// 1.2.1 cross check that ubuntu-boot and ubuntu-seed mounts come from the
+	//       same physical disk, if they don't something went wrong and we need
+	//       to stop booting
+	bootDisk, err := partition.DiskFromMountPoint(boot.InitramfsUbuntuBootDir)
+	if err != nil {
+		return err
+	}
+	seedDisk, err := partition.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir)
+	if err != nil {
+		return err
+	}
+	if !bootDisk.Equals(seedDisk) {
+		// whoops ...
+		return fmt.Errorf("ubuntu-seed partition and ubuntu-boot partition are not from the same disk")
 	}
 
-	// 1.2 mount Data, and exit, as it needs to be mounted for us to do step 2
+	// 1.3 mount Data, and exit, as it needs to be mounted for us to do step 2
 	isDataMounted, err := osutilIsMounted(boot.InitramfsUbuntuDataDir)
 	if err != nil {
 		return err
 	}
 	if !isDataMounted {
 		name := filepath.Base(boot.InitramfsUbuntuDataDir)
-		device, err := unlockIfEncrypted(name)
-		if err != nil {
-			return err
-		}
+		return maybeUnlockAndMount(name)
+	}
 
-		fmt.Fprintf(stdout, "%s %s\n", device, boot.InitramfsUbuntuDataDir)
-		return nil
+	// 1.3.1 cross check that ubuntu-data and ubuntu-boot mounts come from the
+	//       same physical disk, if they don't something went wrong and we need
+	//       to stop booting
+	dataDisk, err := partition.DiskFromMountPoint(boot.InitramfsUbuntuDataDir)
```

Comment:
after having finally booted a vm with encryption and tpm on, I find that this check is wrong, since ubuntu-data will be a device mapper device and hence not be on the same physical disk as the boot disk. I have however talked with @cmatsuoka and refactored this check to actually find that the partition that ubuntu-data comes from, i.e. ubuntu-data-enc, is on the same disk as ubuntu-boot.

Reply:
this check is now fixed